# Initial Data Analysis – trends.csv

This notebook performs an initial analysis of `trends.csv` using Python Pandas. It checks the dataset structure, data types, missing values, duplicates, unique values, and basic statistical measures.

In [1]:
import pandas as pd

df = pd.read_csv("trends.csv")
print("Dataset loaded successfully.")

Dataset loaded successfully.


**Explanation:** The CSV file is loaded into a Pandas DataFrame named `df`.

In [2]:
print(df.head())

  location  year         category  rank  query
0   Global  2001  Consumer Brands     1  Nokia
1   Global  2001  Consumer Brands     2   Sony
2   Global  2001  Consumer Brands     3    BMW
3   Global  2001  Consumer Brands     4   Palm
4   Global  2001  Consumer Brands     5  Adobe


**Explanation:** `head()` displays the first five rows. It is useful for checking the initial records and confirming that the dataset was loaded correctly.

In [3]:
print(df.tail())

      location  year category  rank                query
26950  Vietnam  2020   Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020   Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020   Là Gì?     5       Đông Lào là gì


**Explanation:** `tail()` displays the last five rows. It helps inspect the end of the dataset for unexpected or missing values.

In [4]:
print("Shape:", df.shape)
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Shape: (26955, 5)
Rows: 26955
Columns: 5


**Explanation:** `shape` returns the number of rows and columns. This dataset has 26,955 rows and 5 columns.

In [5]:
print(df.columns.tolist())

['location', 'year', 'category', 'rank', 'query']


**Explanation:** The `columns` attribute lists all variables in the dataset.

In [6]:
print(df.dtypes)

location    object
year         int64
category    object
rank         int64
query       object
dtype: object


**Explanation:** `dtypes` shows the data type of each column. These types determine how the variables can be analyzed and whether conversion may be needed.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26955 entries, 0 to 26954
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   location  26955 non-null  object
 1   year      26955 non-null  int64 
 2   category  26955 non-null  object
 3   rank      26955 non-null  int64 
 4   query     26955 non-null  object
dtypes: int64(2), object(3)
memory usage: 1.0+ MB


**Explanation:** `info()` summarizes the DataFrame, including row count, non-null counts, data types, and memory usage. Comparing non-null counts with the total rows helps identify missing data.

In [8]:
print(df.describe())

               year         rank
count  26955.000000  26955.00000
mean    2015.243369      3.00000
std        3.564683      1.41424
min     2001.000000      1.00000
25%     2013.000000      2.00000
50%     2016.000000      3.00000
75%     2018.000000      4.00000
max     2020.000000      5.00000


**Explanation:** `describe()` gives numerical summary statistics such as count, mean, standard deviation, minimum, quartiles, and maximum for numeric columns.

In [9]:
print(df.describe(include="all").T)

            count unique            top  freq  ...     25%     50%     75%     max
location    26955     83  United States  2070  ...     NaN     NaN     NaN     NaN
year      26955.0    NaN            NaN   NaN  ...  2013.0  2016.0  2018.0  2020.0
category    26955   2450         People   760  ...     NaN     NaN     NaN     NaN
rank      26955.0    NaN            NaN   NaN  ...     2.0     3.0     4.0     5.0
query       26955  18431    Paul Walker    84  ...     NaN     NaN     NaN     NaN

[5 rows x 11 columns]


**Explanation:** `describe(include="all")` extends the summary to categorical/text columns as well. It can show counts, unique values, most frequent values, and frequencies for non-numeric columns.

In [10]:
missing = df.isnull().sum()
print(missing)
print("Total missing values:", missing.sum())

location    0
year        0
category    0
rank        0
query       0
dtype: int64
Total missing values: 0


**Explanation:** `isnull().sum()` counts missing values in every column. Columns with non-zero counts require further cleaning or an appropriate missing-value treatment.

In [11]:
duplicates = df.duplicated().sum()
print("Duplicate rows:", duplicates)

Duplicate rows: 10


**Explanation:** `duplicated().sum()` counts complete duplicate records. Duplicate rows may need to be removed if they are accidental repetitions.

In [12]:
print(df.nunique())

location       83
year           20
category     2450
rank            5
query       18431
dtype: int64


**Explanation:** `nunique()` reports the number of distinct values in each column. It helps identify categorical fields and variables with very low or very high diversity.

In [13]:
for col in df.columns:
    print(f"\n{col}: {df[col].nunique()} unique values")
    print(df[col].dropna().unique()[:10])


location: 83 unique values
['Global' 'France' 'Germany' 'United Kingdom' 'Australia' 'Canada' 'Italy'
 'Netherlands' 'Spain' 'United States']

year: 20 unique values
[2001 2002 2003 2004 2005 2006 2007 2008 2009 2010]

category: 2450 unique values
['Consumer Brands' 'Men' 'Movies' 'MP3 Music Services' 'Music Groups'
 'New Products' 'News Resources' 'Retailers' 'Sports' 'TV Shows']

rank: 5 unique values
[1 2 3 4 5]

query: 18431 unique values
['Nokia' 'Sony' 'BMW' 'Palm' 'Adobe' 'Nostradamus' 'Osama bin Laden'
 'Eminem' 'Michael Jackson' 'Howard Stern']


**Explanation:** This displays the first 10 distinct values from each column as a compact inspection. It avoids printing thousands of search terms while still showing the type of values present.

In [14]:
numeric_cols = df.select_dtypes(include="number").columns
print("Numerical columns:", list(numeric_cols))
print("\nMean:")
print(df[numeric_cols].mean())
print("\nMedian:")
print(df[numeric_cols].median())
print("\nMinimum:")
print(df[numeric_cols].min())
print("\nMaximum:")
print(df[numeric_cols].max())
print("\nStandard deviation:")
print(df[numeric_cols].std())

Numerical columns: ['year', 'rank']

Mean:
year    2015.243369
rank       3.000000
dtype: float64

Median:
year    2016.0
rank       3.0
dtype: float64

Minimum:
year    2001
rank       1
dtype: int64

Maximum:
year    2020
rank       5
dtype: int64

Standard deviation:
year    3.564683
rank    1.414240
dtype: float64


**Explanation:** These calculations provide the requested mean, median, minimum, maximum, and standard deviation for every numerical column.

In [15]:
print("Missing values by column:")
print(df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())
print("\nData types:")
print(df.dtypes)

Missing values by column:
location    0
year        0
category    0
rank        0
query       0
dtype: int64



Duplicate rows: 10

Data types:
location    object
year         int64
category    object
rank         int64
query       object
dtype: object


**Explanation:** This final data-quality check summarizes the main cleaning issues found during the initial analysis.

## Overall Interpretation

The initial analysis provides a structural and quality overview of `trends.csv`. The dataset contains location, year, category, rank, and query information. The numerical variables can be summarized statistically, while the categorical variables can be examined through their unique-value counts and frequency information. Any missing values, duplicate rows, unusual ranges, or inappropriate data types identified in the outputs should be considered during the next data-cleaning stage.